playground/engine.py

```diff
@@ -0,0 +1,818 @@
+    # Glass
+    refractive_index: Optional[float] = None
+    refractive_index_tensor: torch.Tensor = None
+
+    transform: ObjectTransform() = None
+
+    @classmethod
+    def stack(cls, primitives):
+        device = primitives[0].vertices.device
+        vertices = torch.cat([p.vertices for p in primitives], dim=0)
+        v_offset = torch.tensor([0] + [p.vertices.shape[0] for p in primitives[:-1]], device=device)
+        v_offset = torch.cumsum(v_offset, dim=0)
+        triangles = torch.cat([p.triangles + offset for p, offset in zip(primitives, v_offset)], dim=0)
+
+        return OptixPrimitive(
+            vertices=vertices.float(),
+            triangles=triangles.int(),
+            vertex_normals=torch.cat([p.vertex_normals for p in primitives], dim=0).float(),
+            has_tangents=torch.cat([p.has_tangents for p in primitives], dim=0).bool(),
+            vertex_tangents=torch.cat([p.vertex_tangents for p in primitives], dim=0).float(),
+            material_uv=torch.cat([p.material_uv for p in primitives if p.material_uv is not None], dim=0).float(),
+            material_id=torch.cat([p.material_id for p in primitives if p.material_id is not None], dim=0).int(),
+            primitive_type_tensor=torch.cat([p.primitive_type_tensor for p in primitives], dim=0).int(),
+            reflectance_scatter=torch.cat([p.reflectance_scatter for p in primitives], dim=0).float(),
+            refractive_index_tensor=torch.cat([p.refractive_index_tensor for p in primitives], dim=0).float()
+        )
+
+    def apply_transform(self):
+        model_matrix = self.transform.model_matrix()
+        rs_comp = model_matrix[None, :3, :3]
+        t_comp = model_matrix[None, :3, 3:]
+        transformed_verts = (rs_comp @ self.vertices[:, :, None] + t_comp).squeeze(2)
+
+        normal_matrix = self.transform.rotation_matrix()[None, :3, :3]
+        transformed_normals = (normal_matrix @ self.vertex_normals[:, :, None]).squeeze(2)
+        transformed_normals = torch.nn.functional.normalize(transformed_normals)
+
+        transformed_tangents = (normal_matrix @ self.vertex_tangents[:, :, None]).squeeze(2)
+        transformed_tangents = torch.nn.functional.normalize(transformed_tangents)
+
+        return OptixPrimitive(
+            vertices=transformed_verts,
+            triangles=self.triangles,
+            vertex_normals=transformed_normals,
+            vertex_tangents=transformed_tangents,
+            has_tangents=self.has_tangents,
+            material_uv=self.material_uv,
+            material_id=self.material_id,
+            primitive_type=self.primitive_type,
+            primitive_type_tensor=self.primitive_type_tensor,
+            reflectance_scatter=self.reflectance_scatter,
+            refractive_index=self.refractive_index,
+            refractive_index_tensor=self.refractive_index_tensor,
+            transform=ObjectTransform(device=self.transform.device)
+        )
+
+
+class Primitives:
```

Comment:
this class is confusing; hard to tell from name/constructor what it does. Please add a docstring with class mission.